### DISCUSSION TITLE
Driver Syntax error with ODBC SQL Server

### DISCUSSION THREAD

**Discussion by cepbrian** [sqlalchemy/sqlalchemy#8636] [Usage Questions] [answered]
I'm trying to use sqlAlchemy with an AWS instance of SQL server,  setting:

'DRIVER={{ODBC Driver 18 for SQL Server}}'
gives me: 
sqlalchemy.exc.DBAPIError: (pyodbc.Error) ('IM012', '[IM012] [Microsoft][ODBC Driver Manager] DRIVER keyword syntax error (0) (SQLDriverConnect)')

Pretty much any other version of this gives me 
sqlalchemy.exc.InterfaceError: (pyodbc.InterfaceError) ('IM002', '[IM002] [Microsoft][ODBC Driver Manager] Data source name not found and no default driver specified (0) (SQLDriverConnect)')

I'm pretty sure that the { } version is the correct name for the driver, but that seems to be a syntax error.

the {{ }} are because I'm generating a string and formatting into it between when I set the driver and when I generate the connection...

**Comment by zzzeek** [comment, marked as answer]
well that syntax wouldn't be in a SQLAlchemy URL directly can you share what you are actually passing to create_engine.

for examples please see https://docs.sqlalchemy.org/en/14/dialects/mssql.html#connecting-to-pyodbc

**Comment by cepbrian** [comment]
I am calling sqlalchemy.create_engine with:
mssql+pyodbc://P-CEP-SQL/{}?driver={{ODBC Driver 18 for SQL Server}}&authentication=ActiveDirectoryIntegrated

the engine generated prints out as:

Engine(mssql+pyodbc://P-CEP-SQL/services_vars?authentication=ActiveDirectoryIntegrated&driver={ODBC Driver 18 for SQL Server})

**Reply by zzzeek** [reply]
this example is right at https://docs.sqlalchemy.org/en/14/dialects/mssql.html#hostname-connections  

so use this form:

```py
engine = create_engine(
    "mssql+pyodbc://scott:tiger@myhost:49242/databasename"
    "?driver=ODBC+Driver+17+for+SQL+Server"
    "&authentication=ActiveDirectoryIntegrated"
)
```

**Comment by cepbrian** [comment]
Huh, that works for 17, but not 18.... weird. I thought 18 was current.

with 17 it gives me 

sqlalchemy.exc.DBAPIError: (pyodbc.Error) ('FA001', '[FA001] [Microsoft][ODBC Driver 17 for SQL Server]Cannot use Authentication option with Integrated Security option. (0) (SQLDriverConnect); [FA001] [Microsoft][ODBC Driver 17 for SQL Server]Cannot use Authentication option with Integrated Security option. (0)')

but removing the &authentication line works so I'm calling it a win. 

thanks!

**Reply by zzzeek** [reply]
OK what if you use "18" but remove the auth option.   The syntax we are using here should not change.   cc @gordthompson

**Reply by cepbrian** [reply]
18 gives me:
(pyodbc.InterfaceError) ('IM002', '[IM002] [Microsoft][ODBC Driver Manager] Data source name not found and no default driver specified (0) (SQLDriverConnect)')

Which I interpret as "driver not found" because that's what it gives for complete garbage...

**Reply by gordthompson** [reply]
Check the list returned by `pyodbc.drivers()` to see what ODBC drivers are available to your Python app.

**Reply by cepbrian** [reply]
18 doesn't show up in there even though I've installed it.

**Reply by gordthompson** [reply]
I see from your [Stack Overflow question](https://stackoverflow.com/q/74061402/2144390) that you are trying to connect from a Windows machine. What do you see if you run this?

```python
import struct
print("running as {0}-bit".format(struct.calcsize("P") * 8))
```

**Reply by cepbrian** [reply]
running as 64-bit

and I've installed the x64 version. 

I'm not actually sure if this matters, if the 17 version works, my users probably have the 17 version installed not the 18, so it's probably fine, right? My users are all internal to my company, so I CAN force them to upgrade if I need to, but if not necessary, it's not worth the effort...

**Reply by gordthompson** [reply]
It sounds like you have run the x64 installer for "ODBC Driver 17 for SQL Server" but not for "ODBC Driver 18 for SQL Server". They are completely separate installs. You can verify by inspecting the values in

HKEY_LOCAL_MACHINE\SOFTWARE\ODBC\ODBCINST.INI\ODBC Drivers
